File: 0018-4sum/0018-4sum.java
class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
      Arrays.sort(nums);
      int n = nums.length,i,j,k,l;
      List<List<Integer>> result = new ArrayList<>();
      for(i =0;i<n-3;i++)
      {
        if(i>0 && nums[i]==nums[i-1])
          continue;
        for( j = i+1;j<n-2;j++)
        {
            if(j>i+1 && nums[j]==nums[j-1])
              continue;
            k= j+1;
            l = n-1;
            while(k<l)
            {
                long sum = (long)nums[i]+nums[j]+nums[k]+nums[l];
                if(sum==target)
                {
                    result.add(Arrays.asList(nums[i],nums[j],nums[k],nums[l]));
                    while(k<l && nums[k]==nums[k+1])
                        k++;
                    while(k<l && nums[l]==nums[l-1])
                        l--;
                    k++;
                    l--;
                }
                else if(sum<target)
                    k++;
                else
                    l--;
            }
        }
      } 
      return result;
    }
}

File: 0018-4sum/0018-4sum.cpp
class Solution {
//TC = O(N^3) SC =O(1)
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        sort(nums.begin(),nums.end());
        vector<vector<int>> result;
        int low =0, high=0;
        long long sum =0;
        int n = nums.size();
        for(int i =0;i<n-3;i++)
        {
            for(int j =i+1;j<n-2;j++)
            {
                 low =j+1;
                 high =n-1;
                 sum = (long long)target -(long long)(nums[i]+nums[j]);
                 while(low<high)
                 {
                     if(nums[low]+nums[high]==sum)
                     {
                         result.push_back({nums[i],nums[j],nums[low],nums[high]});
                         while(low<high && nums[low]==nums[low+1])
                          low++;
                         while(low<high && nums[high]==nums[high-1])
                          high--;
                         low++;
                         high--;
                     }
                     else if(nums[low]+nums[high]<sum)
                      low++;
                     else
                     high--;
                 }
                while(j<n-2 && nums[j]==nums[j+1])
                j++;

            }
             while(i<n-3 && nums[i]==nums[i+1])
                i++;
        }

        return result;


     }
};